File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_StealCard.java
package com.broll.gainea.server.core.cards.impl.play;

import com.broll.gainea.server.core.cards.Card;
import com.broll.gainea.server.core.player.Player;
import com.broll.gainea.server.core.utils.PlayerUtils;

import java.util.List;
import java.util.stream.Collectors;

public class C_StealCard extends Card {
    public C_StealCard() {
        super(13, "Bekehrung", "Übernehmt eine Aktionskarte von einem beliebigen Spieler");
        setDrawChance(0.5f);
    }

    @Override
    public boolean isPlayable() {
        return !getStealTargets().isEmpty();
    }

    private List<Player> getStealTargets() {
        return PlayerUtils.getOtherPlayers(game, owner).filter(it -> !it.getCardHandler().getCards().isEmpty()).collect(Collectors.toList());
    }

    @Override
    protected void play() {
        Player player = selectHandler.selectPlayer(owner, getStealTargets(), "Welchen Spieler bekehren?");
        List<Card> cards = player.getCardHandler().getCards();
        Card card = cards.get(selectHandler.selectObject("Wählt eine Karte", cards.stream().map(Card::nt).collect(Collectors.toList())));
        player.getCardHandler().discardCard(card);
        owner.getCardHandler().receiveCard(card);
    }

}

File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_StealCard.kt
package com.broll.gainea.server.core.cards.impl.play

import com.broll.gainea.server.core.cards.Card
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.utils.getOtherPlayers
import com.broll.gainea.server.core.utils.sendUpdate

class C_StealCard :
    Card(13, "Bekehrung", "Übernehmt eine Aktionskarte von einem beliebigen Spieler") {
    init {
        drawChance = 0.5f
    }

    override val isPlayable: Boolean
        get() = stealTargets.isNotEmpty()
    private val stealTargets: List<Player>
        get() = game.getOtherPlayers(owner).filter { it.cardHandler.cards.isNotEmpty() }

    override fun play() {
        val player = selectHandler.selectPlayer(owner, stealTargets, "Welchen Spieler bekehren?")
        val cards = player.cardHandler.cards
        val card = cards[selectHandler.selectObject("Wählt eine Karte", cards.map { it.nt() })]
        player.cardHandler.discardCard(card)
        owner.cardHandler.receiveCard(card)
        //update card counts in all clients
        game.sendUpdate(game.nt())
    }
}
